src/com/xilinx/rapidwright/rwroute/RouteNodeInfo.java

```diff
@@ -0,0 +1,180 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.xilinx.rapidwright.rwroute;
+
+import com.xilinx.rapidwright.device.IntentCode;
+import com.xilinx.rapidwright.device.Node;
+import com.xilinx.rapidwright.device.Tile;
+import com.xilinx.rapidwright.device.TileTypeEnum;
+import com.xilinx.rapidwright.device.Wire;
+import com.xilinx.rapidwright.util.Utils;
+
+public class RouteNodeInfo {
+    public final RouteNodeType type;
+    public final short endTileXCoordinate;
+    public final short endTileYCoordinate;
+    public final short length;
+
+    private RouteNodeInfo(RouteNodeType type,
+                          short endTileXCoordinate,
+                          short endTileYCoordinate,
+                          short length) {
+        this.type = type;
+        this.endTileXCoordinate = endTileXCoordinate;
+        this.endTileYCoordinate = endTileYCoordinate;
+        this.length = length;
+    }
+
+    public static RouteNodeInfo get(Node node) {
+        Wire[] wires = node.getAllWiresInNode();
+        Tile baseTile = node.getTile();
+        TileTypeEnum baseType = baseTile.getTileTypeEnum();
+        Tile endTile = null;
+        boolean pinfeedIntoLaguna = false;
+        for (Wire w : wires) {
+            Tile tile = w.getTile();
+            TileTypeEnum tileType = tile.getTileTypeEnum();
+            boolean lagunaTile = false;
+            if (tileType == TileTypeEnum.INT ||
+                    (lagunaTile = Utils.isLaguna(tileType))) {
+                if (!lagunaTile ||
+                        // Only consider a Laguna tile as an end tile if base tile is Laguna too
+                        // (otherwise it's a PINFEED into a Laguna)
+                        Utils.isLaguna(baseType)) {
+                    boolean endTileWasNotNull = (endTile != null);
+                    endTile = tile;
+                    // Break if this is the second INT tile
+                    if (endTileWasNotNull) break;
+                } else {
+                    assert(!Utils.isLaguna(baseType));
+                    pinfeedIntoLaguna = (node.getIntentCode() == IntentCode.NODE_PINFEED);
+                }
+            }
+        }
+        if (endTile == null) {
+            endTile = node.getTile();
+        }
+
+        RouteNodeType type = getType(node, endTile, pinfeedIntoLaguna);
+        short endTileXCoordinate = getEndTileXCoordinate(node, type, (short) endTile.getTileXCoordinate());
+        short endTileYCoordinate = (short) endTile.getTileYCoordinate();
+        short length = getLength(baseTile, type, endTileXCoordinate, endTileYCoordinate);
+
+        return new RouteNodeInfo(type, endTileXCoordinate, endTileYCoordinate, length);
```

Comment:
This seems like it may get called frequently and creating a 'throw-away' object each time `get()` is called may impact performance.  Is there a way we can have a thread-safe static variable that doesn't require `new` on each call?

Reply:
The only thread-safe variable is a [`ThreadLocal`](https://docs.oracle.com/javase/8/docs/api/java/lang/ThreadLocal.html) which I am under the impression requires a hash lookup -- do you know of any others?

I'm aware that there are a lot of create-objects-and-throw-it-away (e.g. `Node.getDownhillNodes()`) operations so I was hoping that the JVM is efficient with this pattern or that it can optimize the construction some way.

Otherwise, in the current implementation of RWRoute, routing node creation is actually single-threaded currently, so writing into a static variable won't be a problem until we need to make it thread safe.

Reply:
Ok, let's leave it for now.  Has performance changed at all based on this PR?